src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1030,34 +1025,98 @@ internal bool TryCalculateSyntaxOffsetOfPositionInInitializer(int position, Synt
             // |<--preceding init len-->|      ^
             //                             position
 
-            int initializersLength = isStatic ? membersAndInitializers.StaticInitializersSyntaxLength : membersAndInitializers.InstanceInitializersSyntaxLength;
-            int distanceFromInitializerStart = position - siblingInitializers[index].Syntax.Span.Start;
+            int initializersLength = getInitializersLength(isStatic ? membersAndInitializers.StaticInitializers : membersAndInitializers.InstanceInitializers);
+            int distanceFromInitializerStart = position - initializer.Syntax.Span.Start;
 
             int distanceFromCtorBody =
                 initializersLength + ctorInitializerLength -
-                (siblingInitializers[index].PrecedingInitializersLength + distanceFromInitializerStart);
+                (precedingLength + distanceFromInitializerStart);
 
             Debug.Assert(distanceFromCtorBody > 0);
 
             // syntax offset 0 is at the start of the ctor body:
             syntaxOffset = -distanceFromCtorBody;
             return true;
-        }
 
-        private static ImmutableArray<FieldOrPropertyInitializer> GetInitializersInSourceTree(SyntaxTree tree, ImmutableArray<ImmutableArray<FieldOrPropertyInitializer>> initializers)
-        {
-            var builder = ArrayBuilder<FieldOrPropertyInitializer>.GetInstance();
-            foreach (var siblingInitializers in initializers)
+            static bool findInitializer(ImmutableArray<ImmutableArray<FieldOrPropertyInitializer>> initializers, int position, SyntaxTree tree,
+                out FieldOrPropertyInitializer found, out int precedingLength)
+            {
+                precedingLength = 0;
+                for (var groupIndex = 0; groupIndex < initializers.Length; groupIndex++)
```

Comment:
It looks like we were using `IndexOfInitializerContainingPosition()` previously, which uses a binary search. Here we're using a linear search. Is this a potential perf issue? (Are we calling this once per initializer?)

Reply:
> It looks like we were using IndexOfInitializerContainingPosition() previously, which uses a binary search. Here we're using a linear search. Is this a potential perf issue? (Are we calling this once per initializer?)

Previously we used ```GetInitializersInSourceTree``` to build a new array by traversing all groups linearly. Then we did ```IndexOfInitializerContainingPosition``` on the new array. I would say the new code is probably better performance-wise.

---
In reply to: [564111608](https://github.com/dotnet/roslyn/pull/50650#discussion_r564111608) [](ancestors = 564111608)

Reply:
If there are `n` initializers and this method is called once per initializer, the perf is `O(n^2)`.

Reply:
@AlekseyTs pointed out offline that the groups are the initializers per type declaration so `n` is likely small, and the previous code was also traversing the groups linearly.

Reply:
Also, this code is mainly called once per closure/lambda. For state machines (in debug mode), it's called once per hoisted local. So I think it's okay.

---
In reply to: [564126934](https://github.com/dotnet/roslyn/pull/50650#discussion_r564126934) [](ancestors = 564126934)